### PR TITLE
add bare minimum implementation of diagnostic analyzer that uses compiler analyzer driver

### PR DESCRIPTION
**Pull request by heejaechang**
added bare minimum implementation of diagnostic analyzer that uses compiler analyzer driver.

literally it is bare minimum implementation. so there is no state or anything to boost performance. it always do everything from scratch.

but, it proves that as long as given interface is implemented, all existing IDE features will continue working. also, since it generates compilation end action diagnostics, those will just work as well. this can be used as a prototype to see what experience will look like once we actually support that scenario.

### CODE REVIEW

**Comment by heejaechang** [comment]
@mavasani @srivatsn @shyamnamboodiripad can you guys take a look?

File: src/Features/Core/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.cs [hunk 1]
```diff
@@ -57,51 +60,150 @@ public override Task NewSolutionSnapshotAsync(Solution solution, CancellationTok
...
+        private async Task<ImmutableArray<DiagnosticData>> GetProjectDiagnosticsAsync(Project project, CancellationToken cancellationToken)
+        {
+            if (project == null)
+            {
+                return ImmutableArray<DiagnosticData>.Empty;
+            }
+
+            var compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+
+            var analyzers = _analyzerManager.CreateDiagnosticAnalyzers(project);
```
Comment (mavasani) [inline src/Features/Core/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.cs]: Should this be instead renamed to "GetDiagnosticAnalyzers" as it isn't creating them for each invocation?

File: src/Features/Core/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.cs [hunk 2]
```diff
@@ -57,51 +60,150 @@ public override Task NewSolutionSnapshotAsync(Solution solution, CancellationTok
...
 
-        public override Task<IEnumerable<DiagnosticData>> GetDiagnosticsForSpanAsync(Document document, TextSpan range, CancellationToken cancellationToken)
+        public override async Task<IEnumerable<DiagnosticData>> GetDiagnosticsForSpanAsync(Document document, TextSpan range, CancellationToken cancellationToken)
         {
-            return SpecializedTasks.EmptyEnumerable<DiagnosticData>();
+            var diagnostics = await GetDiagnosticsAsync(document.Project.Solution, document.Project.Id, document.Id, cancellationToken).ConfigureAwait(false);
+            return diagnostics.Where(d => range.IntersectsWith(d.TextSpan));
+        }
+
+        private async Task<ImmutableArray<DiagnosticData>> GetProjectDiagnosticsAsync(Project project, CancellationToken cancellationToken)
```
Comment (mavasani) [inline src/Features/Core/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.cs]: So basically we come in here for any client that needs any sort of diagnostics right, except cached one? If we ignore perf then I think you fixed lot of bugs in current driver :)

**Comment by mavasani** [comment]
:+1: Great!

File: src/Features/Core/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.cs (hunk 1 above)
Comment (heejaechang) [inline src/Features/Core/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.cs]: not sure what you meant. GetProjectDiagnosticsAsync generate project wide diagnostics.

File: src/Features/Core/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.cs (hunk 2 above)
Comment (heejaechang) [inline src/Features/Core/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.cs]: ha ha. this one is very slow :)

**Comment by srivatsn** [comment]
:+1:

**Pull request merged**